sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,766 @@
+                n_predictors, duration, n_total_leaves))
+            print("{:<32} {:.3f}s".format('Time spent computing histograms:',
+                                          acc_compute_hist_time))
+            print("{:<32} {:.3f}s".format('Time spent finding best splits:',
+                                          acc_find_split_time))
+            print("{:<32} {:.3f}s".format('Time spent applying splits:',
+                                          acc_apply_split_time))
+            print("{:<32} {:.3f}s".format('Time spent predicting:',
+                                          acc_prediction_time))
+
+        self.train_score_ = np.asarray(self.train_score_)
+        self.validation_score_ = np.asarray(self.validation_score_)
+        del self._in_fit  # hard delete so we're sure it can't be used anymore
+        return self
+
+    def _check_early_stopping(self, X_binned_train, y_train,
+                              X_binned_val, y_val):
+        """Check if fitting should be early-stopped.
+
+        Scores are computed on validation data or on training data.
+        """
+
+        self.train_score_.append(
+            self._get_scores(X_binned_train, y_train))
+
+        if self.validation_fraction is not None:
+            self.validation_score_.append(
+                self._get_scores(X_binned_val, y_val))
+            return self._should_stop(self.validation_score_)
+
+        return self._should_stop(self.train_score_)
+
+    def _should_stop(self, scores):
+        """
+        Return True (do early stopping) if the last n scores aren't better
+        than the (n-1)th-to-last score, up to some tolerance.
+        """
+        reference_position = self.n_iter_no_change + 1
+        if len(scores) < reference_position:
+            return False
+
+        # A higher score is always better. Higher tol means that it will be
+        # harder for subsequent iteration to be considered an improvement upon
+        # the reference score, and therefore it is more likely to early stop
+        # because of the lack of significant improvement.
+        tol = 0 if self.tol is None else self.tol
+        reference_score = scores[-reference_position] + tol
+        recent_scores = scores[-reference_position + 1:]
+        recent_improvements = [score > reference_score
+                               for score in recent_scores]
+        return not any(recent_improvements)
+
+    def _get_scores(self, X_binned, y):
+        """Compute scores on data X_binned with target y.
+
+        Scores are computed with a scorer if scoring parameter is not
+        'loss', else with the loss. As higher is always better, we return
+        -loss_value.
+        """
+
+        if self.scoring != 'loss':
+            return self.scorer_(self, X_binned, y)
+
+        # Else, use loss
+        raw_predictions = self._raw_predict(X_binned)
+        return -self.loss_(y, raw_predictions)
```

Comment:
I see.

I'll try to come up with something sensible for the case `scoring='loss'` (default).

Unfortunately I don't think this is possible with scorers

Reply:
Looking at it again makes me wonder why we even need this `X_binned_small_train` subset.

After all, all the predictions for the training data are alreayd computed in `raw_predictions`. We could compute the loss on that, right?

Right now our implementation considers that it is faster to predict for 10 000 samples and then compute the loss on these 10 000 samples rather than directly compute the loss on all the samples (since the prediction for all the samples has already been done before).

I don't think that's true, the prediction step is much more expensive than computing the loss.

Reply:
When using the scorers with the current API we have no choice but to recompute the predictions with the new set of trees, so better use a subset of the training set.

I agree that for the loss we should compute it incrementally both for training and validation, and furthermore we could reuse the training raw predictions used to compute the gradient and the hessian when fitting to compute the training loss.

Reply:
Ok, done in 01ec7d6 :

- use the whole `raw_predictions` array for checking loss on the training data (no subsampling).
- create raw_predictions_val for  checking loss on validation data
- had to separate `_check_early_stopping` into `_check_early_stopping_loss` and `check_early_stopping_scorer` because it was starting to get very messy.
- when checking early stopping with a scorer, the training data is still subsampled

Reply:
Sorry, forgot to push all the changes, should be ok in 7813e96f9